Docs for woocommerce additional token support (#822)
btcpayserver/btcpayserver-doc: docs/WooCommerceAdditionalTokens.md

# WooCommerce Additional Tokens

> With using the additional token configuration you will be able to have separate payment methods for each configured token/currency. This means you can have separate payment methods for BTC, Lightning Network, LTC, ETH (and ERC20 tokens), Liquid assets, ... you name it. This allows you to issue and use [Liquid Assets](https://help.blockstream.com/hc/en-us/articles/900001543146-What-are-Liquid-assets-) as coupons or vouchers, see more details below.)

Reviewer: ```suggestion
With using the additional token configuration you will be able to have separate payment methods for each configured Currency, Asset, Altcoin or Token. This means you can have separate payment methods for BTC, Lightning Network, LTC, ETH (and ERC20 tokens), Liquid assets, ... you name it. This allows you to issue and use [Liquid Assets](https://help.blockstream.com/hc/en-us/articles/900001543146-What-are-Liquid-assets-) as coupons or vouchers, see more details below.)
```

2. **display name**:   
   The visible text for the payment method on checkout

3. **type**:   
   this can be “**payment**” or “**promotion**” [see explanation above](#token-types)

4. **token icon (optional)**:    
   url to token symbol shown during checkout (can be empty but be sure to include the quotes). You can upload the icon in media manager and copy the url or you can use a link to external site or CDN.

:::danger   
**Important:** All the columns text needs to be enclosed by double quotes `"` and separated by semicolon `; each asset has to be put in a new line.
:::

#### Example token config
```
"BTC_OFFCHAIN";"Lightning BTC";"payment";""
"USDt";"USDt (Liquid Theter)";"payment";"https://example.com/wp-content/uploads/2021/01/usdt.png"
"eKr";"eKrona (Liquid Asset)";"promotion";""
```

After saving you will see each asset to be available as payment method. You can enable/disable them like any other payment method. They won’t have any settings itself for now though (everything is configured by the CSV data). But you can use them together with e.g. woocommerce payment plugins to allow discounts for certain payment methods etc.

// todo: add screenshot

> ## Common WooCommerce use-cases with other plugins

Reviewer: ```suggestion
## Common WooCommerce use-cases using the Additionnal Token Support feature
```

### Use-case 1: limit product to a region/shipping zone
Free plugin used: [Country Based Restrictions for WooCommerce](https://wordpress.org/plugins/woo-product-country-base-restrictions/)
After installing and activating the plugin go to a product in “Product data” block there is a new tab “Country restrictions”. You can configure the restrictions you want there.

Example configuration:   
// todo: add screenshot

### Use-case 2: (Promotion) products should have free shipping

> This is doable with woocommerce out of the box (no plugins needed):

Reviewer: ```suggestion
This adds the ability to offer free shipping when a customer pays with a selected Currency, Asset, Altcoin or Token.
This is doable with woocommerce out of the box (no plugins needed):
```

1. In shipping settings add a new shipping class e.g. “free-shipping”
2. On your shipping zones / shipping method config you need to make sure you set the rate to be 0 for that shipping class but also that “cost” is empty or 0. And your “no shipping class cost” is set to the normal rate (using flat-rate as example):   
   // todo: add screenshot
3. In product settings "Product data" block you have a tab "Shipping", there you set the above created "Free-shipping" class and it will taken care during checkout.   
   // todo: add screenshot

### Use-case 3: limit product payment methods

> E.g. allow only a special token payment for promotional product(s)

Reviewer: ```suggestion
E.g. allow only a certain Currency, Asset, Altcoin or Token to be used as payment for promotional product(s)
```

Free plugin used: [Conditional Payments for WooCommerce](https://wordpress.org/plugins/conditional-payments-for-woocommerce/)

This plugin provides a condition rules builder where you can enable/disable available payment methods for products. See example config in the screenshot:   
// todo: add screenshot

### Use-case 4: discount per payment method

> 

Reviewer: ```suggestion
Adds the ability to offer a discount when customer uses a selected Currency, Asset, Altcoin or Token as payment.

```

Free plugin used: [Discounts Per Payment Method for WooCommerce](https://wordpress.org/plugins/woo-payment-discounts/)

In the now available “Discount per Payment” setting in your WooCommerce settings you have a list of all payment methods and can give percentage or fixed discounts.

// todo: add screenshot

### Use-case 5: make sure promotional products can only be purchased exclusively

> This is needed because promotion token payment methods need to override the product price with 1 (per quantity). Otherwise a user could mix normal products with promo products during checkout, which you want to avoid.

Reviewer: ```suggestion
This is needed because payment methods based on a selected Currency, Asset, Altcoin or Token need to override the product price with 1 (per quantity). Otherwise a user could mix normal products with promo products during checkout, which you want to avoid.
```

Author: I will copy this but enrich it a bit with more information to avoid confusion.

*/
function btcpay_check_promotion_product($valid, $product_id, $quantity) {
  $promotion_tag = 'promotion';
  // Check if there are any items in the cart.
  if (!empty($cart_items = WC()->cart->get_cart()) && $valid) {
    // Check if the product is a promotional product and abort.
    if (has_term($promotion_tag, 'product_tag', $product_id)) {
      wc_add_notice( 'Promotional products can only be purchased exclusively, please remove other items from your cart first.', 'error' );
      return false;
    }
    // Also check the case where one has already a promotion product in the
    // cart and also do not allow adding a normal product in that case.
    foreach ($cart_items as $item) {
      if (has_term($promotion_tag, 'product_tag', $item['product_id'])) {
        wc_add_notice( 'Promotional products can only be purchased exclusively, please proceed with checkout or remove the item first.', 'error' );
        return false;
      }
    }
  }
 
  return $valid;
}
add_filter('woocommerce_add_to_cart_validation', 'btcpay_check_promotion_product', 10, 3);
```

### Use-case 6: Limit the checkout of only 1 piece of a product

> 

Reviewer: ```suggestion
Adds the ability to limit the number of a Currency, Asset, Altcoin or Token that can be spent by a customer in a single checkout. 

Useful for coupon-style promotions that are limited to one discount per checkout.

```
